pkg/controller/direct/iam/iambindings.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func mergeBindings(bindings []v1beta1.IAMPolicyBinding) map[iamBindingKey]v1beta1.IAMPolicyBinding {
+	bindingMap := make(map[iamBindingKey]v1beta1.IAMPolicyBinding)
+	for _, a := range bindings {
+		k := getIamBindingKey(a)
+		b, ok := bindingMap[k]
+		if !ok {
+			bindingMap[k] = *a.DeepCopy()
+			continue
+		}
+		b.Members = mergeMembers(b.Members, a.Members)
+		bindingMap[k] = b
+	}
+	return bindingMap
+}
+
+func computeDeletedMembersPerBinding(bindings, lastAppliedBindings []v1beta1.IAMPolicyBinding) []v1beta1.IAMPolicyBinding {
+	res := make([]v1beta1.IAMPolicyBinding, 0)
+	bindingMap := mergeBindings(bindings)
+	lastAppliedBindingMap := mergeBindings(lastAppliedBindings)
+	for k, a := range lastAppliedBindingMap {
+		b, ok := bindingMap[k]
+		if !ok {
+			res = append(res, *a.DeepCopy())
+			continue
+		}
+		removedMembers := computeDeletedMembers(b.Members, a.Members)
+		if len(removedMembers) > 0 {
+			b.Members = removedMembers
+			res = append(res, b)
+		}
+	}
+	return res
+}
+
+func getIamBindingKey(binding v1beta1.IAMPolicyBinding) iamBindingKey {
+	k := iamBindingKey{}
+	k.Role = binding.Role
+	if binding.Condition != nil {
+		k.Condition = *binding.Condition
+	}
+	return k
+}
+
+func removeMembersPerBinding(bindings, deletedBindings []v1beta1.IAMPolicyBinding) []v1beta1.IAMPolicyBinding {
+	bindingMap := mergeBindings(bindings)
+	for _, a := range deletedBindings {
+		k := getIamBindingKey(a)
+		if b, ok := bindingMap[k]; ok {
+			b.Members = removeDeletedMembers(b.Members, a.Members)
+			bindingMap[k] = b
+		}
+	}
+	res := make([]v1beta1.IAMPolicyBinding, 0)
+	for _, b := range bindingMap {
+		if len(b.Members) > 0 {
+			res = append(res, b)
+		}
+	}
+	return res
+}
+
+func removeDeletedMembers(members, deletedMembers []v1beta1.Member) []v1beta1.Member {
+	memberMap := make(map[v1beta1.Member]bool)
+	for _, m := range deletedMembers {
+		memberMap[m] = true
+	}
+	res := make([]v1beta1.Member, 0)
+	for _, m := range members {
+		if _, ok := memberMap[m]; !ok {
+			res = append(res, m)
+		}
+	}
+	return res
+}
+
+func computeDeletedMembers(members, lastAppliedMembers []v1beta1.Member) []v1beta1.Member {
+	memberMap := make(map[v1beta1.Member]bool)
+	res := make([]v1beta1.Member, 0)
+	for _, m := range members {
+		memberMap[m] = true
+	}
+	for _, m := range lastAppliedMembers {
+		if _, ok := memberMap[m]; !ok {
+			res = append(res, m)
+		}
+	}
+	return res
+}
+
+func mergeMembers(memberSlice1, memberSlice2 []v1beta1.Member) []v1beta1.Member {
+	memberMap := make(map[v1beta1.Member]bool)
+	for _, m := range memberSlice1 {
+		memberMap[m] = true
+	}
+	for _, m := range memberSlice2 {
+		memberMap[m] = true
+	}
+	res := make([]v1beta1.Member, 0)
+	for k := range memberMap {
+		res = append(res, k)
+	}
+	sort.Slice(res, func(i, j int) bool {
+		return res[i] < res[j]
+	})
+	return res
+}
+
+func sortBindingSlice(bindings []v1beta1.IAMPolicyBinding) {
+	sort.Slice(bindings, func(i, j int) bool {
```

Comment:
Note: in many situations, the newer [slices.SortFunc](https://pkg.go.dev/slices#SortFunc) function is more ergonomic and runs faster.